pyk/src/pyk/cterm/cterm.py

```diff
@@ -166,7 +166,7 @@ def match_with_constraint(self, cterm: CTerm) -> CSubst | None:
         if subst is None:
             return None
 
-        constraint = self._ml_impl(cterm.constraints, map(subst, self.constraints))
+        constraint = self._ml_impl(cterm.constraints, map(subst, self.constraints))  # <--
```

Comment:
When constructing the `constraint` part of the `CSubst`, an implication is created.

Reply:
This implication, I believe, is what should hold in order for the instantiation to indeed be from a more general to a more specific state, but not what should be added to the state. I think that what should be added to the state is 
```
_ml_impl(subst.constraint, cterm.constraints)
```
which says that if the substitution holds, then the constraints are the more specific constraints.

Reply:
For example, if I am matching 
```
<top> X </top>
#And { true #Equals X >Int 0 }
```
with
```
<top> Y </top>
#And { true #Equals Y >Int 5 }
```
the substitution resulting from configuration matching would be `{ X -> Y }`.

Then, the code we have now would compute
```
{ true #Equals Y >Int 5 } #Implies { true #Equals Y >Int 0 }
```
which is `#Top`, meaning that the instantiation is to a more specific `CTerm`, but this would not transform one `CTerm` into the other.

On the other hand
```
(X #Equals Y)  #Implies { true #Equals Y >Int 5 }
```
would transform one into the other if `apply` first added it and then applied the substitution.

The way `apply` is currently working, it seems to me that it would be enough to add just the constraint: `{ true #Equals Y >Int 5 }` after the substitution has been applied.

Reply:
> The way `apply` is currently working, it seems to me that it would be enough to add just the constraint: `{ true #Equals Y >Int 5 }` after the substitution has been applied.

Indeed, in order for

```python
# When
csubst = cpattern.match_with_constraint(cterm)

# Then
assert csubst.apply(cpattern) == cterm
```

to pass, it'd be sufficient to ensure something like `csubst.constraint == ml_and(cterm.constraints)`.

The problem is, `CSubst.constraint` is public, so before changing it's meaning we need to check all locations where it's used.

Reply:
I found that CTerm doesn't provide `__eq__`. So the correct equivalence may like this:

```python
assert csubst.apply(cpattern).to_dict() == cterm.to_dict()
```

We may provide normalized return for `CTerm.to_dict()`, or provide a logical `__eq__` for CTerm.

Reply:
`CTerm` is a `dataclass`, the `__eq__` method is auto-generated.

Reply:
Thank you, I didn’t notice that.

Reply:
> I think the spec of the `match_with_constraint` is this, is that correct?

We need to inspect all of the call sites to understand how the function is assumed to behave there.

For example, here's how `match_with_constraint` is used in `KCFG` to create a cover edge: https://github.com/runtimeverification/k/blob/0bfb62cf00b71470ed80a42be339d5f87c858e54/pyk/src/pyk/kcfg/kcfg.py#L854-L863

In an `APRProof` though, `csubst` comes from an `APRProofSubsumeResult`: https://github.com/runtimeverification/k/blob/0bfb62cf00b71470ed80a42be339d5f87c858e54/pyk/src/pyk/proof/reachability.py#L193-L194

Whose `csubst` comes from a subsumption check: https://github.com/runtimeverification/k/blob/0bfb62cf00b71470ed80a42be339d5f87c858e54/pyk/src/pyk/proof/reachability.py#L772-L781

Subsumption check: https://github.com/runtimeverification/k/blob/0bfb62cf00b71470ed80a42be339d5f87c858e54/pyk/src/pyk/proof/reachability.py#L738-L748

Which calls `CTermSymbolic.implies`: https://github.com/runtimeverification/k/blob/0bfb62cf00b71470ed80a42be339d5f87c858e54/pyk/src/pyk/cterm/symbolic.py#L191-L199

Here, the constraints come from the `predicate` field from the [RPC response](https://github.com/runtimeverification/haskell-backend/blob/2faf6c1d71de054f4630e3489efb4631a0f320b5/docs/2022-07-18-JSON-RPC-Server-API.md?plain=1#L361). Also, if the implication is not `valid`, the cover edge will not be added in the first place. So my understanding is:
- In `create_cover(csubst=None)`, the created `csubst` represents a _condition_ under which the entailment holds.
- In `create_cover(csubst=csubst)`, the passed `csubst` is expected to represent some sort of _witness_ of entailment.

@PetarMax, what do you think?

If my analysis is correct, these two uses are not consistent with each other. My solution would be to eliminate the `None` case, which gives us one less calls to `match_with_constraint`.

Reply:
Thank you for your analysis! Here’s how I understand it:

Given `subst = t1.match_with_constraint(t2)`, the comment seems to indicate that `subst.apply(t1) == t2`.

https://github.com/runtimeverification/k/blob/0bfb62cf00b71470ed80a42be339d5f87c858e54/pyk/src/pyk/cterm/cterm.py#L163

A cover is an edge where its `csubst` instantiates the target to the source. The code is implemented like this:

https://github.com/runtimeverification/k/blob/0bfb62cf00b71470ed80a42be339d5f87c858e54/pyk/src/pyk/kcfg/kcfg.py#L858

However, I don’t fully understand the result of `cterm_symbolic.implies`, which comes from the koreclient. But if the cover edge's `csubst` has no other implications, I think `csubst.apply(target_cterm) == node_cterm`.

Therefore, I believe `match_with_constraint` should act as a syntactic implication from cterm to self, similar to the semantic one `cterm_symbolic.implies(node.cterm, target_cterm)`.

So, the spec for `match_with_constraint` should be like:

```python
@pytest.mark.parametrize('t1, t2', MATCH_WITH_CONSTRAINT_TEST_DATA, ids=count())
def test_cterm_match_with_constraint(t1: CTerm, t2: CTerm) -> None:
    # When
    c_subst1 = t1.match_with_constraint(t2)
    c_subst2 = t2.match_with_constraint(t1)

    # Then
    if cterm_symbolic.implies(t2, t1):
        assert c_subst1 is not None and cterm_symbolic.simplify(c_subst1.apply(t1)) == cterm_symbolic.simplify(t2)
    if cterm_symbolic.implies(t1, t2):
        assert c_subst2 is not None and cterm_symbolic.simplify(c_subst2.apply(t2)) == cterm_symbolic.simplify(t1)
```

So we shouldn’t implement it this way, because it will always return `#Top`, as `cterm` implies `self`.

https://github.com/runtimeverification/k/blob/0bfb62cf00b71470ed80a42be339d5f87c858e54/pyk/src/pyk/cterm/cterm.py#L169

Reply:
Without modifying the following `apply` method, the `apply` function will remove the existing constraints of the original `CTerm` and won’t apply the `subst` to the constraints.

https://github.com/runtimeverification/k/blob/0bfb62cf00b71470ed80a42be339d5f87c858e54/pyk/src/pyk/cterm/cterm.py#L334-L337

### Running Examples
Given:
```
t1 = CTerm(generated_top(x))
t2 = CTerm(generated_top(y), (mlEqualsTrue(geInt(y, intToken(0))),))
c_subst1 = t1.match_with_constraint(t2)
```
Then:
```
c_subst1 == {x |-> y, y > 0}
```
Given:
```
t1 = X #And { true #Equals X >Int 0 }
t2 = Y #And { true #Equals Y >Int 5 }
c_subst1 = t1.match_with_constraint(t2)
```
Then:
```
c_subst1 == {X |-> Y, Y > 5 /\ Y > 0}
```

### New Algorithm for This

`constraint = t2.constraints + subst(t1.constraints)`

Reply:
I think we can simplify this algorithm further to align with https://github.com/runtimeverification/k/pull/4499#issuecomment-2292758129

`constraint = t2.constraint`

The advantage of this approach is that if `t2 -> t1`, then `t2.constraints + subst(t1.constraints) == t2.constraints`. If not, at least we can recover `t2`, which aligns with the comment’s description.

Reply:
The previous description had some issues. Currently, `CSubst.apply` can only be applied to `CTerm`s with empty constraints; otherwise, it will throw an error. This is due to `CTerm.kast`. Therefore, the current spec is:

@pytest.mark.parametrize('t1, t2', MATCH_WITH_CONSTRAINT_TEST_DATA, ids=count())
def test_cterm_match_with_constraint(t1: CTerm, t2: CTerm) -> None:
    # When
    c_subst1 = t1.match_with_constraint(t2)
    c_subst2 = t2.match_with_constraint(t1)

    # Then
    assert c_subst1 is not None and c_subst1.apply(CTerm(t1.config)) == t2
    assert c_subst2 is not None and c_subst2.apply(CTerm(t2.config)) == t1